File: app/src/main/java/com/yeraygarcia/recipes/RecipeDetailActivity.java
package com.yeraygarcia.recipes;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.content.Intent;
import android.graphics.Rect;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;

import com.yeraygarcia.recipes.database.repository.RecipeDetailRepository;
import com.yeraygarcia.recipes.util.Debug;
import com.yeraygarcia.recipes.viewmodel.RecipeDetailViewModel;
import com.yeraygarcia.recipes.viewmodel.RecipeDetailViewModelFactory;

import java.util.UUID;

public class RecipeDetailActivity extends AppCompatActivity {

    private static final String TAG_FRAGMENT_RECIPE_DETAIL = "tagRecipeDetailFragment";
    private static final String TAG_FRAGMENT_RECIPE_EDIT = "tagEditRecipeFragment";

    private static final String DEFAULT_FRAGMENT_TAG = TAG_FRAGMENT_RECIPE_DETAIL;
    private static final String KEY_FRAGMENT_TAG = "keyFragmentTag";

    private String mCurrentFragment;
    private UUID mRecipeId;

    private RecipeDetailViewModel mViewModel;
    private FloatingActionButton mFabAddToCart;
    private FloatingActionButton mFabEdit;
    private FloatingActionButton mFabSave;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_recipe_detail);

        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);

        Toolbar toolbar = findViewById(R.id.detail_toolbar);
        setSupportActionBar(toolbar);

        mFabAddToCart = findViewById(R.id.fab_add_to_cart);
        mFabEdit = findViewById(R.id.fab_edit_recipe);
        mFabSave = findViewById(R.id.fab_save_recipe);

        mFabEdit.setOnClickListener(editFab -> selectFragment(TAG_FRAGMENT_RECIPE_EDIT));
        mFabSave.setOnClickListener(saveFab -> selectFragment(TAG_FRAGMENT_RECIPE_DETAIL));

        // Show the Up button in the action bar.
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }

        mRecipeId = (UUID) getIntent().getSerializableExtra(RecipeDetailFragment.ARG_RECIPE_ID);

        selectFragment(getFragmentTagFromArguments(savedInstanceState));

        // get ViewModel for recipe id
        RecipeDetailRepository repository = new RecipeDetailRepository(getApplication());
        RecipeDetailViewModelFactory factory = new RecipeDetailViewModelFactory(repository, mRecipeId);
        mViewModel = ViewModelProviders.of(this, factory).get(RecipeDetailViewModel.class);

        // OnClickListener for the Snackbar button
        View.OnClickListener onClickOpenShoppingList = v -> {
            Intent newIntent = new Intent(this, MainActivity.class);
            newIntent.putExtra(MainActivity.EXTRA_FRAGMENT_ID, R.id.navigation_shopping_list);
            Debug.d(RecipeDetailActivity.this, "set Intent." + MainActivity.EXTRA_FRAGMENT_ID + " = " + R.id.navigation_shopping_list);
            startActivity(newIntent);
        };

        mViewModel.isInShoppingList().observe(this, isInShoppingList -> {
            if (isInShoppingList != null && isInShoppingList) {
                mFabAddToCart.setOnClickListener(view -> {
                    mViewModel.removeFromShoppingList(mRecipeId);
                    Snackbar.make(view, R.string.removed_from_shopping_list, Snackbar.LENGTH_LONG)
                            .setAction(R.string.action_open, onClickOpenShoppingList)
                            .show();
                });
                mFabAddToCart.setImageResource(R.drawable.ic_remove_shopping_cart_24px);
            } else {
                mFabAddToCart.setOnClickListener(view -> {
                    mViewModel.addToShoppingList(mRecipeId);
                    Snackbar.make(view, R.string.added_to_shopping_list, Snackbar.LENGTH_LONG)
                            .setAction(R.string.action_open, onClickOpenShoppingList)
                            .show();
                });
                mFabAddToCart.setImageResource(R.drawable.ic_add_shopping_cart_24px);
            }
        });
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString(KEY_FRAGMENT_TAG, mCurrentFragment);
    }

    private String getFragmentTagFromArguments(Bundle savedInstanceState) {

        // get recipe id from savedInstanceState (if not empty)
        if (savedInstanceState != null && savedInstanceState.containsKey(KEY_FRAGMENT_TAG)) {
            return savedInstanceState.getString(KEY_FRAGMENT_TAG, DEFAULT_FRAGMENT_TAG);
        }

        return DEFAULT_FRAGMENT_TAG;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                if (isVisible(TAG_FRAGMENT_RECIPE_EDIT)) {
                    selectFragment(TAG_FRAGMENT_RECIPE_DETAIL);
                } else {
                    navigateUpTo(new Intent(this, MainActivity.class));
                }
                return true;
        }
        return false;
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent event) {
        if (event.getAction() == MotionEvent.ACTION_DOWN) {
            View v = getCurrentFocus();
            if (v instanceof EditText) {
                Rect outRect = new Rect();
                v.getGlobalVisibleRect(outRect);
                if (!outRect.contains((int) event.getRawX(), (int) event.getRawY())) {
                    v.clearFocus();
                    InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
                    if (imm != null) {
                        imm.hideSoftInputFromWindow(v.getWindowToken(), 0);
                    }
                }
            }
        }
        return super.dispatchTouchEvent(event);
    }


    @Override
    public void onBackPressed() {
        if (isVisible(TAG_FRAGMENT_RECIPE_EDIT)) {
            selectFragment(TAG_FRAGMENT_RECIPE_DETAIL);
        } else {
            super.onBackPressed();
        }
    }

    private boolean isVisible(String tag) {
        return getSupportFragmentManager().findFragmentByTag(tag) != null;
    }

    private void selectFragment(String tag) {
        Fragment fragment = getFragmentByTag(tag);
        if (fragment.getTag() == null || !fragment.getTag().equals(mCurrentFragment)) {
            FragmentManager fragmentManager = getSupportFragmentManager();

            if (tag.equals(TAG_FRAGMENT_RECIPE_DETAIL)) {
                fragmentManager.popBackStack(TAG_FRAGMENT_RECIPE_DETAIL, FragmentManager.POP_BACK_STACK_INCLUSIVE);
                fragmentManager
                        .beginTransaction()
                        .replace(R.id.recipe_detail_container, fragment, tag)
                        .commit();
            } else {
                fragmentManager
                        .beginTransaction()
                        .replace(R.id.recipe_detail_container, fragment, tag)
                        .addToBackStack(TAG_FRAGMENT_RECIPE_DETAIL)
                        .commit();
            }

            mCurrentFragment = tag;
        }
        showButtonsByTag(tag);
    }

    private Fragment getFragmentByTag(String tag) {
        Fragment fragment = getSupportFragmentManager().findFragmentByTag(tag);
        if (fragment == null) {
            switch (tag) {
                case TAG_FRAGMENT_RECIPE_EDIT:
                    return EditRecipeFragment.newInstance(mRecipeId);
                case TAG_FRAGMENT_RECIPE_DETAIL:
                    return RecipeDetailFragment.newInstance(mRecipeId);
            }
        }
        return fragment;
    }

    private void showButtonsByTag(String tag) {
        switch (tag) {
            case TAG_FRAGMENT_RECIPE_EDIT:
                showSaveButton();
                break;
            case TAG_FRAGMENT_RECIPE_DETAIL:
            default:
                showCartAndEditButtons();
        }
    }

    private void showCartAndEditButtons() {
        mFabEdit.setVisibility(View.VISIBLE);
        mFabAddToCart.setVisibility(View.VISIBLE);
        mFabSave.setVisibility(View.GONE);
    }

    private void showSaveButton() {
        mFabEdit.setVisibility(View.GONE);
        mFabAddToCart.setVisibility(View.GONE);
        mFabSave.setVisibility(View.VISIBLE);
    }
}

File: app/src/main/java/com/yeraygarcia/recipes/RecipeDetailActivity.kt
package com.yeraygarcia.recipes

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v7.app.AppCompatActivity
import android.view.MenuItem
import android.view.View
import android.view.WindowManager
import com.yeraygarcia.recipes.database.repository.RecipeDetailRepository
import com.yeraygarcia.recipes.viewmodel.RecipeDetailViewModel
import com.yeraygarcia.recipes.viewmodel.RecipeDetailViewModelFactory
import kotlinx.android.synthetic.main.activity_recipe_detail.*
import java.util.*

class RecipeDetailActivity : AppCompatActivity() {

    private var currentFragment: String? = null
    private lateinit var recipeId: UUID
    private lateinit var viewModel: RecipeDetailViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_recipe_detail)

        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)

        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        fabEditRecipe.setOnClickListener { selectFragment(TAG_FRAGMENT_RECIPE_EDIT) }
        fabSaveRecipe.setOnClickListener { selectFragment(TAG_FRAGMENT_RECIPE_DETAIL) }

        recipeId = intent.getSerializableExtra(RecipeDetailFragment.ARG_RECIPE_ID) as UUID

        selectFragment(getFragmentTagFromArguments(savedInstanceState))

        // get ViewModel for recipe id
        val repository = RecipeDetailRepository(application)
        val factory = RecipeDetailViewModelFactory(repository, recipeId)
        viewModel = ViewModelProviders.of(this, factory).get(RecipeDetailViewModel::class.java)

        viewModel.isInShoppingList.observe(this, Observer { isInShoppingList ->
            if (isInShoppingList != null && isInShoppingList) {
                fabAddToCart.setOnClickListener { view ->
                    viewModel.removeFromShoppingList(recipeId)
                    Snackbar.make(view, R.string.removed_from_shopping_list, Snackbar.LENGTH_LONG)
                        .setAction(R.string.action_open) { openShoppingListActivity() }
                        .show()
                }
                fabAddToCart.setImageResource(R.drawable.ic_remove_shopping_cart_24px)
            } else {
                fabAddToCart.setOnClickListener { view ->
                    viewModel.addToShoppingList(recipeId)
                    Snackbar.make(view, R.string.added_to_shopping_list, Snackbar.LENGTH_LONG)
                        .setAction(R.string.action_open) { openShoppingListActivity() }
                        .show()
                }
                fabAddToCart.setImageResource(R.drawable.ic_add_shopping_cart_24px)
            }
        })
    }

    private fun openShoppingListActivity() {
        Intent(this, MainActivity::class.java).apply {
            putExtra(MainActivity.EXTRA_FRAGMENT_ID, R.id.navigationShoppingList)
            startActivity(this)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString(KEY_FRAGMENT_TAG, currentFragment)
    }

    private fun getFragmentTagFromArguments(savedInstanceState: Bundle?): String {
        // get recipe id from savedInstanceState (if not empty)
        return savedInstanceState?.getString(KEY_FRAGMENT_TAG, DEFAULT_FRAGMENT_TAG)
            ?: DEFAULT_FRAGMENT_TAG
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                if (isVisible(TAG_FRAGMENT_RECIPE_EDIT)) {
                    selectFragment(TAG_FRAGMENT_RECIPE_DETAIL)
                } else {
                    navigateUpTo(Intent(this, MainActivity::class.java))
                }
                return true
            }
        }
        return false
    }

//    override fun dispatchTouchEvent(event: MotionEvent): Boolean {
//        if (event.action == MotionEvent.ACTION_DOWN) {
//            val v = currentFocus
//            if (v is EditText) {
//                val outRect = Rect()
//                v.getGlobalVisibleRect(outRect)
//                if (!outRect.contains(event.rawX.toInt(), event.rawY.toInt())) {
//                    v.clearFocus()
//                    val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager?
//                    imm?.hideSoftInputFromWindow(v.windowToken, 0)
//                }
//            }
//        }
//        return super.dispatchTouchEvent(event)
//    }


    override fun onBackPressed() {
        if (isVisible(TAG_FRAGMENT_RECIPE_EDIT)) {
            selectFragment(TAG_FRAGMENT_RECIPE_DETAIL)
        } else {
            super.onBackPressed()
        }
    }

    private fun isVisible(tag: String): Boolean {
        return supportFragmentManager.findFragmentByTag(tag) != null
    }

    private fun selectFragment(tag: String) {
        val fragment = getFragmentByTag(tag)
        if (fragment.tag == null || fragment.tag != currentFragment) {
            if (tag == TAG_FRAGMENT_RECIPE_DETAIL) {
                supportFragmentManager.popBackStack(
                    TAG_FRAGMENT_RECIPE_DETAIL,
                    FragmentManager.POP_BACK_STACK_INCLUSIVE
                )
                supportFragmentManager
                    .beginTransaction()
                    .replace(R.id.recipeDetailContainer, fragment, tag)
                    .commit()
            } else {
                supportFragmentManager
                    .beginTransaction()
                    .replace(R.id.recipeDetailContainer, fragment, tag)
                    .addToBackStack(TAG_FRAGMENT_RECIPE_DETAIL)
                    .commit()
            }

            currentFragment = tag
        }
        showButtonsByTag(tag)
    }

    private fun getFragmentByTag(tag: String): Fragment {
        return supportFragmentManager.findFragmentByTag(tag)
            ?: if (tag == TAG_FRAGMENT_RECIPE_EDIT) {
                EditRecipeFragment.newInstance(recipeId)
            } else {
                RecipeDetailFragment.newInstance(recipeId)
            }
    }

    private fun showButtonsByTag(tag: String) {
        if (tag == TAG_FRAGMENT_RECIPE_EDIT) {
            showSaveButton()
        } else {
            showCartAndEditButtons()
        }
    }

    private fun showCartAndEditButtons() {
        fabEditRecipe.visibility = View.VISIBLE
        fabAddToCart.visibility = View.VISIBLE
        fabSaveRecipe.visibility = View.GONE
    }

    private fun showSaveButton() {
        fabEditRecipe.visibility = View.GONE
        fabAddToCart.visibility = View.GONE
        fabSaveRecipe.visibility = View.VISIBLE
    }

    companion object {
        private const val TAG_FRAGMENT_RECIPE_DETAIL = "tagRecipeDetailFragment"
        private const val TAG_FRAGMENT_RECIPE_EDIT = "tagEditRecipeFragment"
        private const val DEFAULT_FRAGMENT_TAG = TAG_FRAGMENT_RECIPE_DETAIL
        private const val KEY_FRAGMENT_TAG = "keyFragmentTag"
    }
}
